v2/internal/controllers/crd_networking_privateendpoint_test.go

```diff
@@ -0,0 +1,117 @@
+)
+
+func Test_Networking_PrivateEndpoint_CRUD(t *testing.T) {
+	t.Parallel()
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+
+	saKind := storage.StorageAccount_Kind_Spec_BlobStorage
+	sa := newStorageAccount(tc, rg)
+	sa.Spec.Kind = &saKind
+
+	vnet := newVMVirtualNetwork(tc, testcommon.AsOwner(rg))
+	subnet := newVMSubnet(tc, testcommon.AsOwner(vnet))
+
+	tc.CreateResourcesAndWait(sa, vnet, subnet)
+
+	tc.Expect(sa.Status.Id).ToNot(BeNil())
+
+	endpoint := newPrivateEndpoint(tc, rg, sa, subnet)
+
+	tc.CreateResourceAndWait(endpoint)
+
+	tc.Expect(endpoint.Status.Id).ToNot(BeNil())
+	armId := *endpoint.Status.Id
+
+	old := endpoint.DeepCopy()
+	key := "foo"
+	endpoint.Spec.Tags = map[string]string{key: "bar"}
+
+	tc.PatchResourceAndWait(old, endpoint)
+	tc.Expect(endpoint.Status.Tags).To(HaveKey(key))
+
+	tc.RunParallelSubtests(
+		testcommon.Subtest{
+			Name: "Test_DNSZoneGroup_CRUD",
+			Test: func(tc *testcommon.KubePerTestContext) {
+				DNSZoneGroup_CRUD(tc, vnet, endpoint, rg)
+			},
+		},
+	)
+
+	tc.DeleteResourceAndWait(endpoint)
+
+	// Ensure delete
+	exists, retryAfter, err := tc.AzureClient.HeadByID(tc.Ctx, armId, string(network.APIVersion_Value))
+	tc.Expect(err).ToNot(HaveOccurred())
+	tc.Expect(retryAfter).To(BeZero())
+	tc.Expect(exists).To(BeFalse())
+
+}
+
+func DNSZoneGroup_CRUD(tc *testcommon.KubePerTestContext, vnet *v1beta20201101.VirtualNetwork, endpoint *network.PrivateEndpoint, rg *resources.ResourceGroup) {
+	zone := newPrivateDNSZone(tc, "privatelink.blob.core.windows.net", rg)
+	vnetLink := newVirtualNetworkLink(tc, zone, vnet)
+
+	tc.CreateResourcesAndWait(zone, vnetLink)
+
+	dnsZoneGroup := &network.PrivateEndpointsPrivateDnsZoneGroup{
+		ObjectMeta: tc.MakeObjectMeta("dnszonegroup"),
+		Spec: network.PrivateEndpoints_PrivateDnsZoneGroup_Spec{
+			Owner: testcommon.AsOwner(endpoint),
+			PrivateDnsZoneConfigs: []network.PrivateDnsZoneConfig{
+				{
+					Name:                    to.StringPtr("config"),
+					PrivateDnsZoneReference: tc.MakeReferenceFromResource(zone),
+				},
+			},
+		},
+	}
+
+	tc.CreateResourceAndWait(dnsZoneGroup)
+
+	tc.Expect(dnsZoneGroup.Status.Id).ToNot(BeNil())
+
+	tc.DeleteResource(dnsZoneGroup)
+}
+
+func newPrivateEndpoint(tc *testcommon.KubePerTestContext, rg *resources.ResourceGroup, sa *storage.StorageAccount, subnet *v1beta20201101.VirtualNetworksSubnet) *network.PrivateEndpoint {
+	endpoint := &network.PrivateEndpoint{
+		ObjectMeta: tc.MakeObjectMeta("endpoint"),
+		Spec: network.PrivateEndpoint_Spec{
+			Location: tc.AzureRegion,
+			Owner:    testcommon.AsOwner(rg),
+			PrivateLinkServiceConnections: []network.PrivateLinkServiceConnection{
+				{
+					Name:                        to.StringPtr("testEndpoint"),
+					PrivateLinkServiceReference: tc.MakeReferenceFromResource(sa),
+					GroupIds:                    []string{"blob"}, // TODO: This is a bit weird that user has to figure out the group ID(s).
```

Comment:
is this the same thing they need to do in an ARM template?

Reply:
If it is the same as the ARM template story, we may want to have the GroupIds configurable from config maps so that we can automate the workflow.

Reply:
the problem is, GroupIDs are not findable from a resource, it can be funky to get groupIDs for each type of resource into a configMap. For example, for Batch/BatchAccounts, it can be either `batchAccount` or `nodeManagement`, for CosmosDB/DatabaseAccounts it is `Sql`, for ServiceBus/namespaces it'll be `namespace` etc.

Reply:
I am not sure I follow the configMap angle either. I just wanted to confirm that we're at parity with ARM templates or BICEP here, which I think we are. The user needs to (somehow) know what the GroupID is for their resource. I think if that's the same experience users get in the ARM template/BICEP/Terraform world then that's OK we're at parity with how it is in the rest of Azure.